test(boavizt): add tests for client and api

nikhiln64 opened the pull request:
# In this PR
- Added tests around BoaviztAPI module addressing [issue#16](https://github.com/DigitalPebble/spruce/issues/16)
- BoaviztAPIClientTest covers BoaviztAPIClient
- BoaviztAPITest covers BoaviztAPI

```
[INFO] Tests run: 22, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 110.3 s -- in com.digitalpebble.spruce.modules.boavizta.BoaviztAPIClientTest
[INFO] Tests run: 23, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 65.08 s -- in com.digitalpebble.spruce.modules.boavizta.BoaviztAPITest

```

<img width="757" height="157" alt="Screenshot 2025-08-20 at 1 47 51 PM" src="https://github.com/user-attachments/assets/9b6b5a31-9682-484d-8ce7-e1e87aa26b8f" />

jnioche commented on src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:
```diff
@@ -0,0 +1,373 @@
+    void testConstructorWithNullHost() {
+        // The constructor doesn't actually throw NullPointerException for null host
+        // It just stores the null value, which will cause issues later when making HTTP calls
+        BoaviztAPIClient testClient = new BoaviztAPIClient(null);
```
it would be worth changing the behaviour of the constructor so that an error is thrown

jnioche commented on src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:
```diff
@@ -0,0 +1,373 @@
+
+    @Test
+    void testConstructorWithEmptyHost() {
+        BoaviztAPIClient testClient = new BoaviztAPIClient("");
```
same here - it would be worth changing the behaviour of the constructor so that an error is thrown

nikhiln64 replied:
Sure.

jnioche commented on src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:
```diff
@@ -0,0 +1,373 @@
+    }
+
+    @Test
+    void testGetEnergyEstimatesWithValidInstanceType() throws IOException {
```
given that this test is subsumed by the one below, what is the value of keeping it?

jnioche commented on src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:
```diff
@@ -0,0 +1,373 @@
+        // It just passes "null" as a string to the API
+        try {
+            double[] result = client.getEnergyEstimates(Provider.AWS, null);
+            assertNotNull(result);
```
Worth fixing this behaviour as part of this PR?

jnioche commented on src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:
```diff
@@ -0,0 +1,373 @@
+    @Test
+    void testGetEnergyEstimatesWithEmptyInstanceType() throws IOException {
+        try {
+            double[] result = client.getEnergyEstimates(Provider.AWS, "");
```
Worth fixing this behaviour as part of this PR?

jnioche commented on src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:
```diff
@@ -0,0 +1,373 @@
+    @Test
+    void testGetEnergyEstimatesWithWhitespaceInstanceType() throws IOException {
+        try {
+            double[] result = client.getEnergyEstimates(Provider.AWS, "   ");
```
Worth fixing this behaviour as part of this PR?

jnioche commented on src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:
```diff
@@ -0,0 +1,373 @@
+
+    @Test
+    void testGetEnergyEstimatesWithVeryLongInstanceType() throws IOException {
+        String longInstanceType = "t3.micro".repeat(100); // Create a very long instance type
```
Tried a review by Copilot to see what is was like. This suggestion can probably be ignored

jnioche commented on src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:
```diff
@@ -0,0 +1,373 @@
+        // The method doesn't throw NullPointerException for null provider
+        // It just passes null to the API call
+        try {
+            double[] result = client.getEnergyEstimates(null, "t3.micro");
```
Worth fixing this behaviour as part of the PR?

jnioche commented on src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPITest.java:
```diff
@@ -2,26 +2,267 @@
+    }
+
+    @Test
+    void testProcessWithNullOperation() {
```
given that these tests have a similar behaviour and check that the inputs are not changed, they differ only in their inputs. What about having a parameterized test instead?
https://www.baeldung.com/parameterized-tests-junit-5#6-method

jnioche reviewed (commented):
thanks for the PR @nikhiln64 
you tests expose some valid improvements that could be done in the handling of constructors. Would be good to implement these as part of the PR

jnioche reviewed (changes requested):
Thanks @nikhiln64 this is a great addition to the project and a better version of the PR.
I have left a few comments and suggestions.
Thanks again for your contribution

nikhiln64 commented on src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:
```diff
@@ -3,71 +3,96 @@
 
 import static org.junit.jupiter.api.Assertions.*;
 
-public class BoaviztAPIClientTest extends AbstractBoaviztaTest {
```
Real API responses can vary over time, making it difficult to control the returned data. This limits the ability to reliably test error scenarios (e.g., 403 errors, timeouts) and prevents covering edge cases that the real API may not support.

jnioche replied:
I get your point but then why not apply the same approach to the other test class BoaviztAPITest?
Not having to have Docker for testing would be a plus

nikhiln64 commented on src/test/java/com/digitalpebble/spruce/modules/boavizta/BoaviztAPIClientTest.java:
```diff
@@ -0,0 +1,184 @@
+    private BoaviztAPIClient client;
+    private static final String TEST_HOST = "http://localhost:5000";
+
+    @BeforeEach
```
Segregated them. lmk wyt.

jnioche commented:
What would you think about changing AbstractBoaviztaTest so that the container is started once and for all and not for every test? This would reduce the time taken by the tests and reduce the risk of getting errors

nikhiln64 commented:
@jnioche Addressed your comments and here are the improvements after refactoring:
Specific results:
* Previous approach: ~34 seconds for BoaviztAPI tests
* Current approach: 0.092 seconds for BoaviztAPI tests

Total results:


Metric | Before Refactoring | After Refactoring | Improvement
-- | -- | -- | --
BoaviztAPI Tests | ~34s | 0.092s | 370x faster
Total Test Suite | ~40s+ | 3.225s | 12x faster




<img width="726" height="276" alt="Screenshot 2025-08-31 at 8 19 11 AM" src="https://github.com/user-attachments/assets/951ac4f4-7c7c-4139-a657-f6db967a1bf9" />

jnioche commented:
thanks @nikhiln64 
Since you have moved the tests to use the mocks, I don't think we need to keep _AbstractBoavizta*_ classes at all, the dependency on org.testcontainers can also be removed. What do you think?

jnioche commented:
> thanks @nikhiln64 Since you have moved the tests to use the mocks, I don't think we need to keep _AbstractBoavizta*_ classes at all, the dependency on org.testcontainers can also be removed. What do you think?

I will do that separately.

Thanks for this PR @nikhiln64 and your patience with my comments and suggestions, much appreciated!

Have you used Spruce at all?
